java-checks/src/main/java/org/sonar/java/closeresource/ExecutionState.java

```diff
@@ -0,0 +1,269 @@
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypeCastTree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class ExecutionState {
+  @Nullable
+  public ExecutionState parent;
+  private Map<Symbol, CloseableOccurence> closeableOccurenceBySymbol = Maps.newHashMap();
+  private IssuableSubscriptionVisitor check;
+
+  public ExecutionState(Set<Symbol> excludedCloseables, IssuableSubscriptionVisitor check) {
+    this.check = check;
+    for (Symbol symbol : excludedCloseables) {
+      closeableOccurenceBySymbol.put(symbol, CloseableOccurence.IGNORED);
+    }
+  }
+
+  ExecutionState(ExecutionState parent) {
+    this.parent = parent;
+    this.check = parent.check;
+  }
+
+  public ExecutionState merge(ExecutionState executionState) {
+    for (Map.Entry<Symbol, CloseableOccurence> entry : executionState.closeableOccurenceBySymbol.entrySet()) {
+      Symbol symbol = entry.getKey();
+      CloseableOccurence currentOccurence = getCloseableOccurence(symbol);
+      CloseableOccurence occurenceToMerge = entry.getValue();
+      if (currentOccurence != null) {
+        currentOccurence.state = currentOccurence.state.merge(occurenceToMerge.state);
+        closeableOccurenceBySymbol.put(symbol, currentOccurence);
+      } else if (occurenceToMerge.state.isOpen()) {
+        insertIssue(occurenceToMerge.lastAssignment);
+      }
+    }
+    return this;
+  }
+
+  public ExecutionState overrideBy(ExecutionState currentES) {
+    for (Map.Entry<Symbol, CloseableOccurence> entry : currentES.closeableOccurenceBySymbol.entrySet()) {
+      Symbol symbol = entry.getKey();
+      CloseableOccurence occurence = entry.getValue();
+      if (getCloseableOccurence(symbol) != null) {
+        markAs(symbol, occurence.state);
+      } else {
+        closeableOccurenceBySymbol.put(symbol, occurence);
+      }
+    }
+    return this;
+  }
+
+  public ExecutionState restoreParent() {
+    if (parent != null) {
+      insertIssues();
+      return parent.merge(this);
+    }
+    return this;
+  }
+
+  void insertIssues() {
+    for (Tree tree : getUnclosedClosables()) {
+      insertIssue(tree);
+    }
+  }
+
+  void insertIssue(Tree tree) {
+    Type type;
+    if (tree.is(Tree.Kind.VARIABLE)) {
+      type = ((VariableTree) tree).symbol().type();
+    } else {
+      type = ((IdentifierTree) tree).symbol().type();
+    }
+    check.addIssue(tree, "Close this \"" + type.name() + "\"");
+  }
+
+  void addCloseable(Symbol symbol, Tree lastAssignmentTree, @Nullable ExpressionTree assignmentExpression) {
+    CloseableOccurence newOccurence = new CloseableOccurence(lastAssignmentTree, getCloseableStateFromExpression(symbol, assignmentExpression));
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Remove this instantiation of "CloseableOccurence".[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2440)